h2o-genmodel/src/main/java/hex/genmodel/algos/targetencoder/TargetEncoderAsModelProcessor.java

```diff
@@ -0,0 +1,237 @@
+        return new String(cs);
+    }
+
+    public static class TargetEncoderAsCategoricalEncoder implements CategoricalEncoder {
+    
+        private final TargetEncoderMojoModel _teModel;
+        private final EncodingMap _encodings;
+        private final String _columnName;
+        private final int _offsetIndex;
+        private final Map<String, Integer> _domainMap;
+        
+        public TargetEncoderAsCategoricalEncoder(TargetEncoderMojoModel teModel, 
+                                                 EncodingMap encodings, String columnName, 
+                                                 int offsetIndex, String[] domainValues) {
+            _teModel = teModel;
+            _encodings = encodings;
+            _columnName = columnName;
+            _offsetIndex = offsetIndex;
+            _domainMap = new HashMap<>(domainValues.length);
+            for (int j = 0; j < domainValues.length; j++) {
+                _domainMap.put(domainValues[j], j);
+            }
+        }
+    
+        @Override
+        public boolean encodeCatValue(String level, double[] rawData) {
+            Integer category = _domainMap.get(level);
+            if (category == null) return false;
+            _teModel.encodeCategory(rawData, _offsetIndex, _encodings, category);
+            return true;
+        }
+    
+        @Override
+        public void encodeNA(double[] rawData) {
+            _teModel.encodeNA(rawData, _offsetIndex, _encodings, _columnName);
+        }
+        
+    }
+
+    public static class TargetEncoderRowToRawDataConverter extends DefaultRowToRawDataConverter<TargetEncoderMojoModel> {
+        
+        private final TargetEncoderMojoModel _teModel;
+        
+        
+        public TargetEncoderRowToRawDataConverter(TargetEncoderMojoModel teModel,
+                                                  Map<String, Integer> columnToOffsetIdx,
+                                                  Map<Integer, CategoricalEncoder> offsetToEncoder,
+                                                  EasyPredictModelWrapper.ErrorConsumer errorConsumer,
+                                                  EasyPredictModelWrapper.Config config) {
+            // domainMap is ignored here, however we need the TE columns to be enum encoded to be able to apply TE.
+            super(columnToOffsetIdx,
+                  offsetToEncoder,
+                  errorConsumer, 
+                  config);
+            _teModel = teModel;
+        }
+    
+        @Override
+        public double[] convert(RowData data, double[] rawData) throws PredictException {
+            double[] converted = super.convert(data, rawData);
+            if (!_teModel._keepOriginalCategoricalColumns) {
+                for (String teColumn : _teModel._encodingsByCol.keySet()) data.remove(teColumn);
+            }
+            return converted;
+        }
+    }
+
+    public static class VirtualTargetEncodedModel extends GenModel {
```

Comment:
will add some doc and change visibility: this virtual model is used to "trick" further processors and especially `Categoricalencoding` to provide him with a model/frame description that looks as if TE was already applied.